packages/protocol/contracts/common/FeeBurner.sol

```diff
@@ -0,0 +1,425 @@
+    uint256 balanceToBurn = stableToken.balanceOf(address(this));
+
+    if (limitHit(tokenAddress, balanceToBurn)) {
+      // in case the limit is hit, burn the max possible
+      balanceToBurn = currentDayLimit[tokenAddress];
+      emit DailyLimitHit(tokenAddress, balanceToBurn);
+    }
+
+    address exchangeAddress = registry.getAddressForOrDie(stableToken.getExchangeRegistryId());
+
+    IExchange exchange = IExchange(exchangeAddress);
+
+    uint256 minAmount = 0;
+    if (FixidityLib.unwrap(maxSlippage[tokenAddress]) != 0) {
+      // max slippage is set
+      // use sorted oracles as reference
+      ISortedOracles sortedOracles = getSortedOracles();
+      (uint256 priceWithoutSlippage, ) = sortedOracles.medianRate(tokenAddress);
+      minAmount = calculateMinAmount(priceWithoutSlippage, tokenAddress, balanceToBurn);
+    }
+
+    // small numbers cause rounding errors and zero case should be skipped
+    if (balanceToBurn <= MIN_BURN) {
+      return;
+    }
+
+    // TODO maybe we could compare with uniswap as well
+    stableToken.approve(exchangeAddress, balanceToBurn);
+    exchange.sell(balanceToBurn, minAmount, false);
+    pastBurn[tokenAddress] += balanceToBurn;
+
+    updateLimits(tokenAddress, balanceToBurn);
+
+    emit SoldAndBurnedToken(tokenAddress, balanceToBurn);
+  }
+
+  /**
+    * @notice Burns the max possible of a token.
+    * @dev Should be used in case the loop fails because one token is reverting or OOG.
+    * @param tokenAddress the address of the token to burn.
+    */
+  function burnSingleNonMentoToken(address tokenAddress) public onlyWhenNotFrozen {
+    // An improvement to this function would be to allow the user to pass a path as argument
+    // and if it generates a better outcome that the ones enabled that gets used
+    // and the user gets a reward
+
+    address celoAddress = address(getGoldToken());
+
+    uint256 bestRouterIndex = 0;
+    uint256 bestRouterQuote = 0;
+
+    address[] memory path = new address[](2);
+    address[] memory thisTokenRouterAddresses = routerAddresses[tokenAddress];
+
+    IERC20 token = IERC20(tokenAddress);
+    uint256 balanceToBurn = token.balanceOf(address(this));
+
+    if (limitHit(tokenAddress, balanceToBurn)) {
+      // in case the limit is hit, burn the max possible
+      balanceToBurn = currentDayLimit[tokenAddress];
+      emit DailyLimitHit(tokenAddress, balanceToBurn);
+    }
+
+    // small numbers cause rounding errors and zero case should be skipped
+    if (balanceToBurn <= MIN_BURN) {
+      return;
+    }
+
+    require(thisTokenRouterAddresses.length > 0, "routerAddresses should be non empty");
+
+    for (uint256 j = 0; j < thisTokenRouterAddresses.length; j++) {
+      address poolAddress = thisTokenRouterAddresses[j];
+
+      require(poolAddress != address(0), "poolAddress should be nonZero");
+      IUniswapV2RouterMin router = IUniswapV2RouterMin(poolAddress);
+
+      path[0] = tokenAddress;
```

Comment:
Im wondering if the tokens to be burned will always have pool with Celo.

Reply:
That is why pools are enabled per token. Worst case we just get a zero quote.

Reply:
Why only limit to direct Celo exchange as opposed to having more than 2 path? Wouldn't that allow for better slippage mitigation if low liquidity in token:celo pair?

Reply:
The problem is that a priori, it's difficult to know which one is the best a apriori, so a smart contract'd have to solve a big graph to get the optimal. And as the exchanges are likely arbitraged, using the direct path is likely very close to ideal.

An idea for the future would be to allow a user to submit a path better than the canonical, and then get a reward for doing so.

Reply:
Chatted with @soloseng offline, and we could add more paths for a token. This could be a follow up improvement.